test/integration/protocol_integration_test.cc

```diff
@@ -410,6 +410,71 @@ TEST_P(ProtocolIntegrationTest, RetryStreamingCancelDueToBufferOverflow) {
   test_server_->waitForCounterEq("cluster.cluster_0.retry_or_shadow_abandoned", 1);
 }
 
+bool paused(IntegrationTestServerPtr& server) {
+  if (!server->counter("cluster.cluster_0.upstream_flow_control_backed_up_total") ||
+      !server->counter("cluster.cluster_0.upstream_flow_control_drained_total")) {
+    return false;
+  }
+
+  return server->counter("cluster.cluster_0.upstream_flow_control_backed_up_total")->value() !=
+         server->counter("cluster.cluster_0.upstream_flow_control_drained_total")->value();
+}
+
+TEST_P(ProtocolIntegrationTest, RetryDuringOverrun) {
+  // Set a tiny upstream buffer limit and a very large downstream limit, to
+  // make it easier for the request to back up in the router filter.
+  config_helper_.setBufferLimits(1024, 200000000);
+  initialize();
+
+  // Send request headers upstream.
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  auto headers = Http::TestRequestHeaderMapImpl{
+      {":method", "POST"},    {":path", "/test/long/url"},     {":scheme", "http"},
+      {":authority", "host"}, {"x-forwarded-for", "10.0.0.1"}, {"x-envoy-retry-on", "5xx"}};
+  auto encoder_decoder = codec_client_->startRequest(headers);
+  auto response = std::move(encoder_decoder.second);
+  ASSERT_TRUE(fake_upstreams_[0]->waitForHttpConnection(*dispatcher_, fake_upstream_connection_));
+  ASSERT_TRUE(fake_upstream_connection_->waitForNewStream(*dispatcher_, upstream_request_));
+  ASSERT_TRUE(upstream_request_->waitForHeadersComplete());
+
+  // Read disable, and send a large body until the router backs up.
+  ASSERT_TRUE(fake_upstream_connection_->readDisable(true));
+  do {
+    Buffer::OwnedImpl data(std::string(640000, 'a'));
+    codec_client_->sendData(encoder_decoder.first, data, false);
+    if (paused(test_server_)) {
+      usleep(5000);
```

Comment:
Instead of sleeping, are there any metrics we can monitor (rx bytes on downstream, tx bytes on upstream maybe?) to monitor for when the desired state is achieved?

Reply:
Unfortunately I think it's a similar race there.  I think if I stopped doing this for both H1 and H2 I could probably get something hard-coded that worked at least on a given platform.  Given this is testing in-Envoy retry logic and is pretty protocol agnostic I think that'd be sufficient for an integration test.  Would you prefer that?

Reply:
I'm always a fan of not using sleep in tests whenever possible. If it looks viable, and it'll be less racy than sleep, I'd prefer that. If it ends up being just as racy, don't bother.